File: packages/typescript/src/resources/inbound.ts
/**
 * The `inbound` resource: messages received on your numbers.
 * @module
 */
import { toWire } from '../models.js';
import type { InboundMessage, InboundReplyParams, ListParams, Message, Page, RequestOptions } from '../models.js';
import { Resource, seg, wireQuery } from './base.js';

/** Accessed as `client.inbound`. */
export class Inbound extends Resource {
  list(params: ListParams = {}, options?: RequestOptions): Promise<Page<InboundMessage>> {
    return this.page({ method: 'GET', path: '/v1/inbound', query: wireQuery(params) }, options);
  }

  /** Reply to an inbound message (live keys only). Returns the sent Message. */
  reply(id: string, params: InboundReplyParams, options?: RequestOptions): Promise<Message> {
    return this.one(
      { method: 'POST', path: `/v1/inbound/${seg(id)}/reply`, json: toWire(params), idempotent: true },
      options,
    );
  }
}

File: packages/rust/src/resources/inbound.rs
//! The `inbound` resource: received SMS and replies.

use super::{idem, list_query, seg};
use crate::error::Result;
use crate::http::HttpClient;
use crate::models::{InboundMessage, ListParams, Message, Page, ReplyInbound, RequestOptions};

/// Accessed as `client.inbound()`.
#[derive(Debug, Clone)]
pub struct Inbound {
    http: HttpClient,
}

impl Inbound {
    pub(crate) fn new(http: HttpClient) -> Self {
        Self { http }
    }

    /// List inbound messages.
    pub async fn list(&self, params: ListParams) -> Result<Page<InboundMessage>> {
        self.http
            .get(&format!("/v1/inbound{}", list_query(&params)))
            .await
    }

    /// Reply to an inbound message (live keys only). Returns the sent Message.
    pub async fn reply(&self, id: &str, params: &ReplyInbound) -> Result<Message> {
        self.reply_with(id, params, &RequestOptions::default())
            .await
    }

    /// [`Inbound::reply`] with an explicit Idempotency-Key.
    pub async fn reply_with(
        &self,
        id: &str,
        params: &ReplyInbound,
        opts: &RequestOptions,
    ) -> Result<Message> {
        let path = format!("/v1/inbound/{}/reply", seg(id, "id")?);
        self.http.send("POST", &path, params, idem(opts)).await
    }
}
